add metric for file_sd file timestamps

cstyan opened the pull request:
addresses #2757 
forked from: https://github.com/prometheus/prometheus/pull/3063 and rebased off master

Not sure if this is the proper way to handle Briand's comment here: https://github.com/prometheus/prometheus/pull/3063#issuecomment-323953612

Obviously some changes are needed as `MustRegister` panics if the config is reloaded (the collector is already registered), and if we use just `Register` the metrics disappear on config reload.

brian-brazil commented on discovery/file/file.go:
```diff
@@ -35,6 +38,60 @@ import (
+			"prometheus_sd_file_timestamp",
+			"Timestamp of files read by FileSD",
+			[]string{"filename"},
+			prometheus.Labels{"file_sd_index": strconv.Itoa(i)},
```
The index is an implementation detail, it shouldn't really be exposed to the user.

The challenge here really is how to deal with files that appear in multiple distinct scrape_configs.

cstyan replied:
shouldn't the same file being in multiple file_sd or scrape_config jobs be invalid? but I guess in either case the solution is to have a single map and dedup entry's when we update the map

brian-brazil replied:
It's perfectly valid to do that.

cstyan replied:
so just to confirm, something like this is valid and needs to be handled properly by the timestamp collection code:
```
scrape_configs:
- job_name: "overwritten-default"
  scrape_interval: 10s

  file_sd_configs:
  - files:
    - filesd1.yml
    - filesd2.yml
  - files:
    - filesd2.yml
    - filesd3.yml
- job_name: "jobTwo"
  scrape_interval: 10s
  
  file_sd_configs:
    - filesd1.yml
    - filesd3.yml
```

brian-brazil replied:
Yes

brian-brazil commented on discovery/file/file.go:
```diff
@@ -35,6 +37,71 @@ import (
+	files := t.filenames
+	t.lock.RUnlock()
+	for k := range files {
+		info, err := os.Stat(k)
```
We really should have the mtime at the last time we read it, not just what's currently on the filesystem. This would allow for detecting things being stuck.

brian-brazil commented on discovery/file/file.go:
```diff
@@ -73,6 +142,7 @@ func NewDiscovery(conf *config.FileSDConfig, logger log.Logger) *Discovery {
 	if logger == nil {
 		logger = log.NewNopLogger()
 	}
+	fileSDTimeStamp.logger = logger
```
A constructor shouldn't really be setting a field on a global. I don't think it should be doing any logging.

cstyan replied:
I left this in as I thought we might want to have logging if for some reason we can't access the file to get the timestamp, but that should be caught by this block in `refresh`:
```
tgroups, err := readFile(p)
		if err != nil {
			fileSDReadErrorsCount.Inc()

			level.Error(d.logger).Log("msg", "Error reading file", "path", p, "err", err)
			// Prevent deletion down below.
			ref[p] = d.lastRefresh[p]
			continue
		}
```

brian-brazil commented on discovery/file/file.go:
```diff
@@ -219,6 +289,9 @@ func (d *Discovery) refresh(ctx context.Context, ch chan<- []*config.TargetGroup
 		if !ok || n > m {
+			level.Error(d.logger).Log("msg", "file_sd refresh found file that should be removed", "file", f)
+			// Remove from file sd timestamp collector.
+			fileSDTimeStamp.DeleteFile(f)
```
This feels racy.

I think it'd be better to iterate over the active file SDs, each of which has its own state.

cstyan replied:
this actually isn't needed to get the metric to not be reported if the file is removed, I don't quite understand what you mean by iterating over all active file SDs, that would mean we need to keep track of the timestamps in discovery.go wouldn't it?

cstyan replied:
how about just clearing the map (or assigning a newly created one) to the timestamp collector at the start of refresh? since we always clear all files, and then compare the ones we knew about before to the ones we know about know for updating the target groups, we can just clear the map and then add timestamps for the ones that we're calling `readfile` on in `refresh`

cstyan commented:
@brian-brazil I think my most recent commit addresses most of your comments, though I don't think it entirely solves the 'this feels racy' comment regarding removal of a file's from the collection. The timestamps are now stored as maps by file sd index (we don't expose the index in the metrics though).

The issue is that for multiple scrape configs we have the same index for each file sd, meaning with a config of:
```
scrape_configs:
- job_name: "jobOne"
  scrape_interval: 10s

  file_sd_configs:
  - files:
    - filesd1.yml
    - filesd2.yml
  - files:
    - filesd2.yml
    - filesd3.yml
- job_name: "jobTwo"
  scrape_interval: 10s
  
  file_sd_configs:
  - files:
    - filesd4.yml
```

The timestamp collection ends up looking like this:
```
map[0:map[filesd1.yml:1.512175031e+09 filesd2.yml:1.512173506e+09 filesd4.yml:1.512360496e+09] 1:map[filesd3.yml:1.51217364e+09 filesd2.yml:1.512173506e+09]
```
So `filesd4.yml`'s timestamp is stored in the same map as `filesd1.yml` and `filesd2.yml` from jobOne. Do you have any ideas on how to solve this?

cstyan commented:
also, I think the timestamp collectors functions do not need to be exporter, let me know if I'm wrong here

brian-brazil commented on discovery/file/file.go:
```diff
@@ -240,12 +317,21 @@ func fileSource(filename string, i int) string {
 		return nil, err
 	}
 
+	// We use a mutex when setting timestamp, so don't block the rest of readFile.
```
You'd be better off reading the file using the `os` package and calling `File.Stat`, that way you know it' the timestamp of the file you read.

cstyan replied:
`ioutil.readFile` uses `os.Open`, so I think we're safe here

https://golang.org/src/io/ioutil/ioutil.go?s=1464:1510#L39

cstyan replied:
ah yeah you're right, I see what you mean now

brian-brazil commented on discovery/discovery.go:
```diff
@@ -63,7 +63,7 @@ func ProvidersFromConfig(cfg config.ServiceDiscoveryConfig, logger log.Logger) m
 	}
 	for i, c := range cfg.FileSDConfigs {
-		app("file", i, file.NewDiscovery(c, log.With(logger, "discovery", "file")))
+		app("file", i, file.NewDiscovery(c, log.With(logger, "discovery", "file"), i))
```
This is only the index within one scrape_config, which may change across reloads if the user reorders things.

What I'd suggest doing is having each file discovery register with TimestampCollector. Each file discovery would also remember the mtimes of its files. Then on Collect the TimestampCollector would pull in all this information, and de-dupe.

brian-brazil commented on discovery/file/file.go:
```diff
@@ -73,11 +123,15 @@ func NewDiscovery(conf *config.FileSDConfig, logger log.Logger) *Discovery {
+		timestamps: make(map[string]float64),
+		logger:     logger,
 	}
+	fileSDTimeStamp.discoverers[disc] = struct{}{}
```
You need a mutex for this write.

brian-brazil commented on discovery/file/file.go:
```diff
@@ -162,6 +216,18 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+}
+
 // stop shuts down the file watcher.
 func (d *Discovery) stop() {
```
It should remove itself from the collector in here.

brian-brazil commented on discovery/file/file.go:
```diff
@@ -233,19 +317,28 @@ func (d *Discovery) refresh(ctx context.Context, ch chan<- []*config.TargetGroup
+		return nil, err
+	}
+
+	// We use a mutex when setting timestamp, so don't block the rest of readFile.
```
There's no need for a goroutine her, a single syscall isn't that expensive.

cstyan replied:
This was more for the writing of the timestamp to `Discovery`'s map, I thought if there were a lot of files we wouldn't want to block the reading of each one on getting and storing the timestamp. I'll remove it if you still think it's unnecessary.

brian-brazil replied:
Updating a map is cheaper than creating a goroutine.

brian-brazil commented on discovery/file/file.go:
```diff
@@ -35,6 +37,65 @@ import (
+	return &TimestampCollector{
+		Description: prometheus.NewDesc(
+			"prometheus_sd_file_timestamp",
+			"Timestamp of files read by FileSD. Timestamp is set at read time.",
```
I'd mention that this is the mtime

brian-brazil commented on discovery/file/file.go:
```diff
@@ -219,6 +301,8 @@ func (d *Discovery) refresh(ctx context.Context, ch chan<- []*config.TargetGroup
 	for f, n := range d.lastRefresh {
 		m, ok := ref[f]
 		if !ok || n > m {
+			level.Error(d.logger).Log("msg", "file_sd refresh found file that should be removed", "file", f)
```
Is this really an error?

cstyan replied:
no :)

brian-brazil commented on discovery/file/file.go:
```diff
@@ -233,18 +317,24 @@ func (d *Discovery) refresh(ctx context.Context, ch chan<- []*config.TargetGroup
-func readFile(filename string) ([]*config.TargetGroup, error) {
-	content, err := ioutil.ReadFile(filename)
+func (d *Discovery) readFile(filename string) ([]*config.TargetGroup, error) {
+	fd, err := os.Open(filename)
```
You need to close this file descriptor

brian-brazil commented on discovery/file/file.go:
```diff
@@ -233,18 +317,24 @@ func (d *Discovery) refresh(ctx context.Context, ch chan<- []*config.TargetGroup
 	if err != nil {
 		return nil, err
 	}
+	d.writeTimestamp(filename, float64(info.ModTime().Unix()))
```
This would be best down the end, as if there's a failure to parse the file this metric getting stuck would help indicate that.

brian-brazil commented on discovery/file/file.go:
```diff
@@ -35,6 +37,65 @@ import (
+
+// Collect creates constant metrics for each file with last modified time of the file.
+func (t *TimestampCollector) Collect(ch chan<- prometheus.Metric) {
+	// new map to dedup dilenames
```
Start comments with a capital letter and end with a full stop.

brian-brazil commented on discovery/file/file.go:
```diff
@@ -35,6 +37,65 @@ import (
+	// new map to dedup dilenames
+	uniqueFiles := make(map[string]float64)
+	t.lock.RLock()
+	discoverers := t.discoverers
```
This just copies a reference, you need to hold the lock until after the first for loop.

cstyan commented:
@brian-brazil thanks for the continued reviews, appreciate it :)

brian-brazil commented on discovery/file/file.go:
```diff
@@ -233,15 +316,21 @@ func (d *Discovery) refresh(ctx context.Context, ch chan<- []*config.TargetGroup
-	content, err := ioutil.ReadFile(filename)
+func (d *Discovery) readFile(filename string) ([]*config.TargetGroup, error) {
+	fd, err := os.Open(filename)
+	defer fd.Close()
```
This should go after the error check, otherwise fd could be nil.

brian-brazil commented:
Can you squash can your commits, leaving the original commit on its own to preserve authorship?

cstyan commented:
bump @brian-brazil

brian-brazil commented:
Thanks!
